source/iterator/snapshot.go

```diff
@@ -0,0 +1,322 @@
+	EntityType       config.EntityType
+	EntityLabels     []string
+	BatchSize        int
+	DatabaseName     string
+	Position         sdk.Position
+}
+
+// NewSnapshot creates a new instance of the [Snapshot].
+func NewSnapshot(ctx context.Context, params SnapshotParams) (*Snapshot, error) {
+	position, err := parsePosition(params.Position)
+	if err != nil && !errors.Is(err, errNilSDKPosition) {
+		return nil, fmt.Errorf("parse position: %w", err)
+	}
+
+	var (
+		orderingPropertyMaxValue any
+		// join entity labels here to not do this for each individual element
+		entityLabels = strings.Join(params.EntityLabels, ":")
+	)
+	switch pos := position; {
+	case pos != nil && position.MaxElement != nil:
+		orderingPropertyMaxValue = position.MaxElement
+
+	default:
+		orderingPropertyMaxValue, err = getMaxPropertyValue(
+			ctx, params.Driver,
+			params.DatabaseName, entityLabels, params.OrderingProperty,
+			params.EntityType,
+		)
+		if err != nil && !errors.Is(err, errNoElements) {
+			return nil, fmt.Errorf("get ordering property max value: %w", err)
+		}
+	}
+
+	return &Snapshot{
+		driver:                   params.Driver,
+		keyProperties:            params.KeyProperties,
+		orderingProperty:         params.OrderingProperty,
+		orderingPropertyMaxValue: orderingPropertyMaxValue,
+		entityType:               params.EntityType,
+		entityLabels:             entityLabels,
+		batchSize:                params.BatchSize,
+		databaseName:             params.DatabaseName,
+		position:                 position,
+		records:                  make(chan map[string]any, params.BatchSize),
+	}, nil
+}
+
+// HasNext checks whether the snapshot iterator has records to return or not.
+func (s *Snapshot) HasNext(ctx context.Context) (bool, error) {
+	if len(s.records) > 0 {
+		return true, nil
+	}
+
+	if err := s.loadBatch(ctx); err != nil {
+		return false, fmt.Errorf("load batch: %w", err)
+	}
+
+	return len(s.records) > 0, nil
+}
+
+// Next returns the next available record.
+func (s *Snapshot) Next(ctx context.Context) (sdk.Record, error) {
+	select {
+	case <-ctx.Done():
+		return sdk.Record{}, ctx.Err() //nolint:wrapcheck // there's no much to wrap here
+
+	case record := <-s.records:
+		// construct the position
+		position := &position{
+			Mode:       modeSnapshot,
+			Element:    record[s.orderingProperty],
+			MaxElement: s.orderingPropertyMaxValue,
+		}
+
+		sdkPosition, err := position.marshalSDKPosition()
+		if err != nil {
+			return sdk.Record{}, fmt.Errorf("marshal sdk position: %w", err)
+		}
+
+		s.position = position
+
+		// construct the key
+		key := make(sdk.StructuredData)
+		for _, keyProperty := range s.keyProperties {
+			keyPropertyValue, ok := record[keyProperty]
+			if !ok {
+				return sdk.Record{}, fmt.Errorf("payload doesn't contain %q property", keyProperty)
+			}
+
+			key[keyProperty] = keyPropertyValue
+		}
+
+		// construct the metadata
+		metadata := make(sdk.Metadata)
+		metadata.SetCreatedAt(time.Now())
```

Comment:
What do you think to add entity name to metadata?

Reply:
Yeah, sounds great, I've added a new constant `metadataEntityLabelsField` for this purpose.